aten/src/ATen/core/dispatch/OperatorEntry.cpp

```diff
@@ -153,49 +153,71 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
   //          Now that we first call Autograd backend keys after BackendSelect, we should fill those
   //          with catchAll as well.
   //  (3) Use fallthrough kernel that are registered as fallback.
   //  (4) Use catchAll kernel if available
-  // TODO: currently Autograd is the only alias key, we'll update alias key precedence after we add new
-  //      alias keys AutogradDispatchCPUOrCUDA and Math.
-  // TODO: we can fix (2.2) and remove (4) after TypeDefault registrations are moved from catchAll to Math
+  // Alias Key Precedence:
+  //   Math > Autograd
+  // TODO: Update alias key precedence after we add new alias keys AutogradDispatchCPUOrCUDA .
+  // TODO: we can remove (2.3) and (4) after TypeDefault registrations are moved from catchAll to Math
   //       so that Math can populate to Autograd backend keys before fallback kernels.
 
   // 1. Operator registration
-  auto kern_it = kernels_.find(dispatch_key);
-  if (kern_it != kernels_.end()) {
-    TORCH_INTERNAL_ASSERT(!kern_it->second.empty());
-    TORCH_INTERNAL_ASSERT(kern_it->second.front().kernel.isValid());
-    return {kern_it->second.front(), "kernel"};
-
-  } else if (isIncludedInAlias(dispatch_key, DispatchKey::Autograd)) {
-    // 2.1. For autograd backend keys, use kernel from DispatchKey::Autograd if available
-    auto kern_autograd = kernels_.find(DispatchKey::Autograd);
-    if (kern_autograd != kernels_.end()) {
-      TORCH_INTERNAL_ASSERT(!kern_autograd->second.empty());
-      TORCH_INTERNAL_ASSERT(kern_autograd->second.front().kernel.isValid());
-      return {kern_autograd->second.front(), "autograd kernel"};
-
-    // 2.2. For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
-    //      if there's no direct registration to the backend key.
-    } else if (kernels_.find(getBackendKeyFromAutograd(dispatch_key)) == kernels_.end()
-            && !catchAllKernel_.empty()) {
-      TORCH_INTERNAL_ASSERT(catchAllKernel_.front().kernel.isValid());
-      return {catchAllKernel_.front(), "catch all"};
-    }
+  if (hasKernelForDispatchKey(dispatch_key)) {
+    return {getKernelForDispatchKey(dispatch_key), "kernel"};
+  }
+
+  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
+  //      when there's no direct registration to its corresponding backend key.
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Math)
+      && hasKernelForDispatchKey(DispatchKey::Math)
+      && !(isIncludedInAlias(dispatch_key, DispatchKey::Autograd)
+           && hasKernelForDispatchKey(getBackendKeyFromAutograd(dispatch_key)))) {
+    return {getKernelForDispatchKey(DispatchKey::Math), "math kernel"};
+  }
+
+  // 2.2. For autograd backend keys, use kernel from DispatchKey::Autograd if available
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Autograd)
+      && hasKernelForDispatchKey(DispatchKey::Autograd)) {
+    return {getKernelForDispatchKey(DispatchKey::Autograd), "autograd kernel"};
```

Comment:
Kind of an aside, but I'm mildly averse to the "hasThing/getThing" idiom for these helpers bc they repeat the linear search. NBD in precomputation, and the code is definitely neater with the lookup factored... but in a perfect world I'd probably have the helper return an optional, just in case code cargo-culted from here winds up in a hotspot. 

Eh I dunno if it's worth tearing up the sidewalk. Really I'm still thinking about the AutogradOther thing 😁

Reply:
Looks like we need to return an optional reference here. Just to double check, do we prefer ` optional<std::reference_wrapper<T>>` or a raw pointer if we want to do that? :P

Reply:
The benefit of reference wrapper is that it is guaranteed to be non-null. cc @smessmer